[Shipping labels] Improve address validation errors

irfano opened the pull request:
<!-- Remember about a good descriptive title. -->

<!-- Id number of the Linear issue this PR addresses, e.g., WOOMOB-373. -->
Closes WOOMOB-904

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
This PR adds support for displaying address field errors on the Edit Address screen, using the validation errors returned from the endpoint. For example, when the house number is missing, the API response looks like this:
```
 "errors": {
      "address": "House number is missing",
      "general": "Address not found"
    },
```

### Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->
1. Go to Orders.
2. Select an order that is eligible for creating shipping labels.
3. Tap the Create shipping label button.
4. Tap the Shipment details button to open the bottom sheet.
5. Tap the Edit button next to "Ship to".
6. Modify the address and enter an invalid one (e.g., omit the house number).
7. Tap "Validate & Save".

### Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->
Repeat the steps for the origin address as well, but ensure that unverified addresses can't be saved as origin addresses.

### The tests that have been performed
<!-- To give the reviewer an idea of what could be missed in terms of testing -->
Steps above

### Images/gif
<!-- Include before and after images or gifs when appropriate. -->
|Before|After|
|-|-|
|<img width="1344" height="2992" alt="Screenshot_20250814_000502" src="https://github.com/user-attachments/assets/09c0adb2-8051-4d3b-bae5-1549e71764a2" />|<img width="1344" height="2992" alt="Screenshot_20250813_235844" src="https://github.com/user-attachments/assets/a2dac74d-a076-422f-b9e8-2be7cde6c01f" />|

- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

irfano committed:
Make NormalizationResponseDTO parameters nullable

irfano committed:
Add errors to address normalization model

This commit updates the `AddressNormalizationModel` to include a new `errors` field.
The `WooShippingNetworkingMapper` is also updated to map the errors from the DTO to the model.

irfano committed:
Handle errors parameter for normalizeAddress in the repository

irfano committed:
Refactor NormalizeAddress error constants

irfano committed:
Refactor NormalizeAddressException to support multiple errors

The `NormalizeAddressException` class is updated to accept a map of errors instead of a single error string. This allows for more granular error reporting, specifically differentiating between general errors and address-specific errors.

irfano committed:
Refactor address normalization error handling

The error handling in `NormalizeAddress.kt` is updated to provide more specific error messages.
- If the response model is null, an "Empty response" message is used.
- `NormalizeAddressException` is now used when there are specific address errors.
- A general `Exception` is used for other errors, incorporating the error message from the response if available.

irfano committed:
Refactor AddressStatus to be a sealed class

The `AddressStatus` enum has been converted to a sealed class. This change allows for more detailed state representation, particularly for the `VerifyFailed` status, which can now optionally include a `NormalizeAddressException`.

irfano committed:
Improve address normalization error handling and analytics

irfano committed:
Update address with normalize exception

irfano committed:
Update address status according to normalize exception

irfano committed:
Update getErrorState() logic to handle new normalize exception

irfano committed:
Shorten WooShippingEditAddressViewModel.kt class

dangermattic commented:
<!--
  0 Errors
  2 Warnings: This PR is larger than 300 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-android/milestone/265">23.1</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

irfano committed:
Update RELEASE-NOTES.txt

irfano commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/wooshippinglabels/address/EditAddressViewState.kt:
I created this class to address the `LargeClass` Detekt error in the `WooShippingEditAddressViewModel` class.

irfano commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/wooshippinglabels/address/NormalizeAddress.kt:
```diff
@@ -12,22 +12,30 @@ class NormalizeAddress @Inject constructor(
     private val site: SelectedSite,
 ) {
     suspend operator fun invoke(address: Address): Result<AddressNormalizationModel> {
-        return site.getOrNull()?.let {
```
I changed `site.getOrNull()` to `site.get()` to align with the approach we use for our domain classes in the shipping label screens.

irfano committed:
Merge branch 'trunk' into issue/WOOMOB-904-improve-address-validation-errors

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/wooshippinglabels/address/NormalizeAddress.kt:
```diff
@@ -12,22 +12,30 @@ class NormalizeAddress @Inject constructor(
 
-class NormalizeAddressException(val error: String) : Exception(error) {
+class NormalizeAddressException(val errors: Map<String, String>) : Exception() {
+    val generalError: String?
+        get() = errors[ERROR_GENERAL]
+    val addressError: String?
+        get() = errors[ERROR_ADDRESS]
```
Just a remark here, the iOS app exposes an error for the `name` field ([here](https://github.com/woocommerce/woocommerce-ios/blob/047ccd83f0ed1c2fa67adeb3fcaef95641253e9c/WooCommerce/Classes/ViewRelated/Orders/Order%20Details/Shipping%20Labels/WooShipping%20Create%20Shipping%20Labels/WooShippingAddresses/WooShippingEditAddressViewModel.swift#L394-L402)), and the plugin takes this a step further by accepting an error for all fields (Check lines 43 to 49 of the file `address-step/fields.jsx`).

I'm not sure how to trigger any of the other errors, so I'm not sure if we need to support this or not, WDYT?

irfano replied:
I didn't add errors for the other fields because I couldn't reproduce them. I reviewed `address-step/fields.jsx` but still couldn't determine the mapping. In this PR, I mapped `address` key to the "Address" field, but I couldn't find the keys for the other fields.
@itsmeichigo, since you used `name` for the "Name" field errors, have you found a way to reproduce the `name` error?

hichamboushaba replied:
> I reviewed address-step/fields.jsx but still couldn't determine the mapping

You can find all the keys by looking for `getProps` on the file: `name`, `company`, `email`, `phone`, `country`, `address`, `city`, `state` and `postcode`. The mapping with of the error with the `key` is done on line 43.

> I didn't add errors for the other fields because I couldn't reproduce them

IMO we can add them without being able to reproduce them, in case the errors happen for some specific countries, but that's now a blocker for merging this PR.

irfano replied:
Thanks for your input, @hichamboushaba! I’ve added errors for `name`, `company`, `email`, `phone`, `city`, and `postcode`. I think it’s worth testing again, **could you test it once more?**

<img width="400" alt="Screenshot_20250819_202713" src="https://github.com/user-attachments/assets/d8040116-cafa-472e-b8a3-f3d9c0ae3ab6" />

I couldn’t test the errors with API Faker, so I used the Network Inspector’s Rules tab with the following mocked response:

```
{
  "data": {
    "success": false,
    "errors": {
      "name": "Testing name error",
      "company": "Testing company error",
      "address": "House number is missing",
      "city": "Testing city error",
      "state": "Testing state error",
      "postcode": "Testing postcode error",
      "email": "Testing email error",
      "phone": "Testing phone error",
      "general": "Address not found"
    },
    "isTrivialNormalization": false,
    "address": {
      "address_1": "60 29TH STREET #343",
      "address_2": "",
      "city": "SAN FRANCISCO",
      "company": "",
      "country": "US",
      "default_address": false,
      "email": "irfanotest@gmail.com",
      "is_verified": false,
      "name": "IRFAN OMUR",
      "phone": "",
      "postcode": "94110-4929",
      "state": "CA"
    }
  }
}
```

I didn't add test for `country` and `state`. Country field is always dropdown, so it's not necessary to add error for it. `state` can be editable for some cases but I decided not to add error for it as well since it's rarely not a dropdown, effort to add error for state field was larger and I didn't want to increase the scope of this PR. Plus we aren't even sure server send errors for states. Let me know if you think we should add it.

hichamboushaba replied:
Thanks @irfano, I tested the changes using ApiFaker, and it works as expected.

> I didn't add test for country and state. Country field is always dropdown, so it's not necessary to add error for it. state can be editable for some cases but I decided not to add error for it as well since it's rarely not a dropdown, effort to add error for state field was larger and I didn't want to increase the scope of this PR. Plus we aren't even sure server send errors for states. Let me know if you think we should add it.

Sounds good to me.

hichamboushaba reviewed (approved):
Nice work @irfano, works well.
I just left a question, but I'm pre-approving.

irfano committed:
Merge branch 'trunk' into issue/WOOMOB-904-improve-address-validation-errors

irfano committed:
Display individual field errors for address normalization API

Previously, only a general address error was shown. Now, errors for name, company, address, city, postcode, email, and phone are surfaced individually.

irfano committed:
Merge branch 'trunk' into issue/WOOMOB-904-improve-address-validation-errors

irfano committed:
Merge branch 'trunk' into issue/WOOMOB-904-improve-address-validation-errors

irfano committed:
Merge branch 'trunk' into issue/WOOMOB-904-improve-address-validation-errors
